File: lessons/dsa/queue-and-first-in-order/examples/FirstInOrderTest.java
import java.util.List;

public final class FirstInOrderTest {
  private static int failures = 0;
  private static final List<String> LESSON =
      List.of("+report.pdf", "+photo.jpg", "-", "+notes.txt", "-", "-");

  private static void check(String name, boolean passed) {
    if (!passed) {
      System.out.println("FAIL " + name);
      failures++;
    }
  }

  public static void main(String[] args) {
    // The defining property, over several event sequences rather than one.
    List<List<String>> sequences =
        List.of(
            List.of("+a", "-"),
            LESSON,
            List.of("+a", "+b", "+c", "-", "-", "-"),
            List.of("+a", "-", "+b", "-", "+c", "-"));
    for (List<String> events : sequences) {
      check("served in arrival order",
          FirstInOrder.serveFirstIn(events).equals(FirstInOrder.arrivalsIn(events)));
    }

    check("the lesson order",
        FirstInOrder.serveFirstIn(LESSON)
            .equals(List.of("report.pdf", "photo.jpg", "notes.txt")));

    // The contrast the two lessons exist to draw, on identical input.
    check("a pile serves a different order",
        FirstInOrder.serveLastIn(LESSON)
            .equals(List.of("photo.jpg", "notes.txt", "report.pdf")));

    // notes.txt arrives after photo.jpg is already waiting and is served after
    // it. This is the step where the pile does the opposite.
    List<String> served = FirstInOrder.serveFirstIn(LESSON);
    List<String> piled = FirstInOrder.serveLastIn(LESSON);
    check("a late arrival does not overtake",
        served.indexOf("photo.jpg") < served.indexOf("notes.txt"));
    check("the pile serves the newer first",
        piled.indexOf("notes.txt") < piled.indexOf("report.pdf"));

    // Not "may be delayed": under alternating arrivals and services a pile
    // never serves the first item at all, for any number of rounds.
    for (int rounds : new int[] {1, 5, 50}) {
      List<String> events = FirstInOrder.steadyStream("first", rounds);
      check("a queue serves the first arrival",
          FirstInOrder.serveFirstIn(events).contains("first"));
      check("a pile starves it",
          !FirstInOrder.serveLastIn(events).contains("first"));
    }

    check("the queue serves it immediately",
        FirstInOrder.serveFirstIn(FirstInOrder.steadyStream("first", 5))
            .get(0)
            .equals("first"));

    check("serving an empty line does nothing",
        FirstInOrder.serveFirstIn(List.of("-", "-")).isEmpty());
    check("serving after an empty line still works",
        FirstInOrder.serveFirstIn(List.of("-", "+a", "-")).equals(List.of("a")));
    check("unserved arrivals are not served",
        FirstInOrder.serveFirstIn(List.of("+a", "+b", "-")).equals(List.of("a")));
    check("no events serve nobody",
        FirstInOrder.serveFirstIn(List.of()).isEmpty()
            && FirstInOrder.serveLastIn(List.of()).isEmpty());
    check("repeated names are ordinary",
        FirstInOrder.serveFirstIn(List.of("+job", "+job", "-", "-"))
            .equals(List.of("job", "job")));

    if (failures > 0) {
      System.exit(1);
    }
    System.out.println("All checks passed.");
  }
}

File: lessons/dsa/queue-and-first-in-order/examples/first_in_order_test.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include "first_in_order.cpp"

namespace {
int failures = 0;
const std::vector<std::string> kLesson = {"+report.pdf", "+photo.jpg", "-",
                                          "+notes.txt", "-", "-"};

void check(const char* name, bool passed) {
  if (!passed) {
    std::cout << "FAIL " << name << "\n";
    ++failures;
  }
}

bool holds(const std::vector<std::string>& items, const std::string& name) {
  return std::find(items.begin(), items.end(), name) != items.end();
}

long positionOf(const std::vector<std::string>& items, const std::string& name) {
  return std::find(items.begin(), items.end(), name) - items.begin();
}
}  // namespace

int main() {
  // The defining property, over several event sequences rather than one.
  const std::vector<std::vector<std::string>> sequences = {
      {"+a", "-"},
      kLesson,
      {"+a", "+b", "+c", "-", "-", "-"},
      {"+a", "-", "+b", "-", "+c", "-"}};
  for (const std::vector<std::string>& events : sequences) {
    check("served in arrival order", serveFirstIn(events) == arrivalsIn(events));
  }

  check("the lesson order",
        serveFirstIn(kLesson) ==
            std::vector<std::string>({"report.pdf", "photo.jpg", "notes.txt"}));

  // The contrast the two lessons exist to draw, on identical input.
  check("a pile serves a different order",
        serveLastIn(kLesson) ==
            std::vector<std::string>({"photo.jpg", "notes.txt", "report.pdf"}));

  // notes.txt arrives after photo.jpg is already waiting and is served after
  // it. This is the step where the pile does the opposite.
  const std::vector<std::string> served = serveFirstIn(kLesson);
  const std::vector<std::string> piled = serveLastIn(kLesson);
  check("a late arrival does not overtake",
        positionOf(served, "photo.jpg") < positionOf(served, "notes.txt"));
  check("the pile serves the newer first",
        positionOf(piled, "notes.txt") < positionOf(piled, "report.pdf"));

  // Not "may be delayed": under alternating arrivals and services a pile never
  // serves the first item at all, for any number of rounds.
  for (int rounds : {1, 5, 50}) {
    const std::vector<std::string> events = steadyStream("first", rounds);
    check("a queue serves the first arrival", holds(serveFirstIn(events), "first"));
    check("a pile starves it", !holds(serveLastIn(events), "first"));
  }

  check("the queue serves it immediately",
        serveFirstIn(steadyStream("first", 5)).front() == "first");

  check("serving an empty line does nothing", serveFirstIn({"-", "-"}).empty());
  check("serving after an empty line still works",
        serveFirstIn({"-", "+a", "-"}) == std::vector<std::string>({"a"}));
  check("unserved arrivals are not served",
        serveFirstIn({"+a", "+b", "-"}) == std::vector<std::string>({"a"}));
  check("no events serve nobody",
        serveFirstIn({}).empty() && serveLastIn({}).empty());
  check("repeated names are ordinary",
        serveFirstIn({"+job", "+job", "-", "-"}) ==
            std::vector<std::string>({"job", "job"}));

  if (failures > 0) {
    return EXIT_FAILURE;
  }
  std::cout << "All checks passed.\n";
  return EXIT_SUCCESS;
}
